kong/clustering/services/sync/strategies/postgres.lua

```diff
@@ -0,0 +1,122 @@
+local _M = {}
+local _MT = { __index = _M }
+
+
+local cjson = require("cjson.safe")
+local buffer = require("string.buffer")
+
+
+local string_format = string.format
+local cjson_encode = cjson.encode
+local ngx_log = ngx.log
+local ngx_ERR = ngx.ERR
+local ngx_DEBUG = ngx.DEBUG
+
+
+local CLEANUP_VERSION_COUNT = 100
+local CLEANUP_TIME_DELAY = 3600  -- 1 hour
+
+
+function _M.new(db)
+  local self = {
+    connector = db.connector,
+  }
+
+  return setmetatable(self, _MT)
+end
+
+
+local PURGE_QUERY = [[
+  DELETE FROM clustering_sync_version
+  WHERE "version" < (
+      SELECT MAX("version") - %d
+      FROM clustering_sync_version
+  );
+]]
+
+
+function _M:init_worker()
+  ngx.timer.every(CLEANUP_TIME_DELAY, function(premature)
+    if premature then
+      ngx_log(ngx_DEBUG, "[incremental] worker exiting, killing incremental cleanup timer")
+
+      return
+    end
+
+    local res, err = self.connector:query(string_format(PURGE_QUERY, CLEANUP_VERSION_COUNT))
+    if not res then
+      ngx_log(ngx_ERR,
+              "[incremental] unable to purge old data from incremental delta table, err: ",
+              err)
+
+      return
+    end
+
+    ngx_log(ngx_DEBUG,
+            "[incremental] successfully purged old data from incremental delta table")
+  end)
+end
+
+
+local NEW_VERSION_QUERY = [[
+  DO $$
+  DECLARE
+    new_version integer;
+  BEGIN
+    INSERT INTO clustering_sync_version DEFAULT VALUES RETURNING version INTO new_version;
+    INSERT INTO clustering_sync_delta (version, type, id, ws_id, row) VALUES %s;
+  END $$;
+]]
+
+
+-- deltas: {
+--   { type = "service", "id" = "d78eb00f-8702-4d6a-bfd9-e005f904ae3e", "ws_id" = "73478cf6-964f-412d-b1c4-8ac88d9e85e9", row = "JSON", }
+--   { type = "route", "id" = "0a5bac5c-b795-4981-95d2-919ba3390b7e", "ws_id" = "73478cf6-964f-412d-b1c4-8ac88d9e85e9", row = "JSON", }
+-- }
+function _M:insert_delta(deltas)
+  local buf = buffer.new()
+  for _, d in ipairs(deltas) do
+    buf:putf("(new_version, %s, %s, %s, %s)",
+             self.connector:escape_literal(d.type),
+             self.connector:escape_literal(d.id),
+             self.connector:escape_literal(d.ws_id),
+             self.connector:escape_literal(cjson_encode(d.row)))
+  end
+
+  local sql = string_format(NEW_VERSION_QUERY, buf:get())
+
+  return self.connector:query(sql)
+end
+
+
+function _M:get_latest_version()
+  local sql = "SELECT MAX(version) AS max_version FROM clustering_sync_version"
+  return self.connector:query(sql)[1].max_version
```

Comment:
The other methods in this file all return directly from `connector:query()`, so I assume they will return `nil, string` in case of error.

On the other hand, this method may raise an error due to invalid table access if there is a failure in `connector:query()`.

I suggest we make the error-handling more consistent:

```suggestion
  local rows, err = self.connector:query(sql)
  if not rows then
    return nil, err
  end
  
  -- should we raise/return an error if no rows/max_version are returned?
  -- or should we return a default value of 0?
  return rows[1] and rows[1].max_version
```

If it's intended to raise an error when `connector:query()` fails, then I suggest using `assert()` or `error()` to make the intention more explicit.

We should also codify the behavior in the case that `clustering_sync_version` is empty. I'm seeing code elsewhere that assumes it cannot return `nil`. Is the caller incorrect?

```lua
-- location: kong/clustering/services/sync/rpc.lua

function _M:init_cp(manager)
-- [...]
    if default_namespace_version == 0 or
       self.strategy:get_latest_version() - default_namespace_version > FULL_SYNC_THRESHOLD
    then
-- [...]
```

Reply:
fixed in https://github.com/Kong/kong/pull/13157/commits/2838ffd5c7ca9323bfa888ada11834ecec3b6911